File: backend/src/schemas/jobAnalysis.ts
import { z } from "zod";

/**
 * Shape produced by Step 1 (Perplexity or manual).
 */
export const TagsForGraphQLSchema = z.object({
  company: z.string().min(1).optional(),
  roleFamily: z.string().min(1).optional(),
  seniority: z.string().min(1).optional(),
  domains: z.array(z.string().min(1)).default([]),
  stack: z.array(z.string().min(1)).default([]),
  kinds: z.array(z.enum(["experience", "education", "project"])).default(["experience", "education", "project"]),
  groups: z.array(z.string().min(1)).default([])
});

export type TagsForGraphQL = z.infer<typeof TagsForGraphQLSchema>;

export const JobAnalysisSchema = z.object({
  schemaVersion: z.literal(1),
  generatedAt: z.string().min(1),
  jobSource: z.string().min(1),
  tagsForGraphQL: TagsForGraphQLSchema,
  tagsForTokenization: z.array(z.string().min(1)).default([]),
  summaryForGeneration: z.string().min(1),
  rawProviderResponse: z.string().optional()
});

export type JobAnalysis = z.infer<typeof JobAnalysisSchema>;

File: backend/src/resume_orchestrator/schemas/job_analysis.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

Kind = Literal["experience", "education", "project"]


class TagsForGraphQL(BaseModel):
    model_config = ConfigDict(extra="ignore")

    company: str | None = None
    roleFamily: str | None = None
    seniority: str | None = None
    domains: list[str] = Field(default_factory=list)
    stack: list[str] = Field(default_factory=list)
    kinds: list[Kind] = Field(
        default_factory=lambda: ["experience", "education", "project"]
    )
    groups: list[str] = Field(default_factory=list)


class JobAnalysis(BaseModel):
    model_config = ConfigDict(extra="ignore")

    schemaVersion: Literal[1] = 1
    generatedAt: str = Field(min_length=1)
    jobSource: str = Field(min_length=1)
    tagsForGraphQL: TagsForGraphQL
    tagsForTokenization: list[str] = Field(default_factory=list)
    summaryForGeneration: str = Field(min_length=1)
    rawProviderResponse: str | None = None
